craft_application/models/project.py

```diff
@@ -93,14 +96,81 @@ class BuildPlannerConfig(CraftBaseConfig):
     """The BuildPlanner model uses attributes from the project yaml."""
 
 
+class Platform(CraftBaseModel):
+    """Project platform definition."""
+
+    build_on: list[str] | None = pydantic.Field(min_items=1, unique_items=True)
+    build_for: list[str] | None = pydantic.Field(
+        min_items=1, max_items=1, unique_items=True
+    )
+
+    @pydantic.root_validator(  # pyright: ignore[reportUntypedFunctionDecorator,reportUnknownMemberType]
+        skip_on_failure=True
+    )
+    @classmethod
+    def _validate_platform_set(cls, values: Mapping[str, Any]) -> Mapping[str, Any]:
+        """If build_for is provided, then build_on must also be."""
+        if not values.get("build_on") and values.get("build_for"):
+            raise errors.CraftValidationError(
+                "'build_for' expects 'build_on' to also be provided."
+            )
+
+        return values
+
+
 class BuildPlanner(CraftBaseModel, metaclass=abc.ABCMeta):
     """The BuildPlanner obtains a build plan for the project."""
 
+    platforms: dict[str, Platform | None] = {
```

Comment:
A shared pre-validator in the `BuildPlanner` and `Project` models should set a default value.

```suggestion
    platforms: dict[str, Platform] = {
```